cli/link.ts

```diff
@@ -0,0 +1,65 @@
+import type {ArgumentsCamelCase} from 'yargs';
+import {createCodaClient} from './helpers';
+import {formatEndpoint} from './helpers';
+import {getApiKey} from './config_storage';
+import {getPackId} from './config_storage';
+import {printAndExit} from '../testing/helpers';
+import {promptForInput} from '../testing/helpers';
+import {storePackId} from './config_storage';
+
+// Regular expression that matches coda.io/p/<packId> or <packId>.
+const PackUrlRegex = /^(?:https:\/\/(?:[^/]*)coda.io(?:\:[0-9]+)?\/p\/)?([0-9]+)(:?[^0-9]*)$/;
+
+interface LinkArgs {
+  manifestDir: string;
+  codaApiEndpoint: string;
+  packIdOrUrl: string;
+}
+
+export function parsePackIdOrUrl(packIdOrUrl: string): number | null {
+  const match = packIdOrUrl.match(PackUrlRegex);
+  if (!match) {
+    return null;
+  }
+  const matchedNumber = match[1];
+  return Number(matchedNumber);
+}
+
+export async function handleLink({manifestDir, codaApiEndpoint, packIdOrUrl}: ArgumentsCamelCase<LinkArgs>) {
+  const formattedEndpoint = formatEndpoint(codaApiEndpoint);
+  // TODO(dweitzman): Redirect to the `coda register` if there's not
+  // an existing Coda API token.
+  // TODO(dweitzman): Add a download command to fetch the latest code from
+  // the server and ask people if they want to download after linking.
+  const apiKey = getApiKey(codaApiEndpoint);
+  if (!apiKey) {
+    printAndExit('Missing API key. Please run `coda register <apiKey>` to register one.');
+  }
+
+  const codaClient = createCodaClient(apiKey, formattedEndpoint);
+  const packId = parsePackIdOrUrl(packIdOrUrl);
+  if (packId === null) {
+    printAndExit(`packIdOrUrl must be a pack ID or URL, not ${packIdOrUrl}`);
```

Comment:
Huh, TypeScript must have recently gotten smarter about inferring type narrowing in cases like this. You used to need to add a `return` here otherwise the code below wouldn't have been properly narrowed to know that `packId` is non-null. I think that since `printAndExit()` is typed as returning `never`, a recent version of TS is able to narrow the type even without a return.

Regardless, I'd recommend using `return` in all of these cases to make it clearer that these lines are terminal.